crates/ruff_linter/src/rules/pylint/rules/missing_maxsplit_arg.rs

```diff
@@ -0,0 +1,146 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_python_ast::{
+    DictItem, Expr, ExprAttribute, ExprCall, ExprDict, ExprNumberLiteral, ExprStringLiteral,
+    ExprSubscript, ExprUnaryOp, Keyword, Number, UnaryOp,
+};
+use ruff_python_semantic::{analyze::typing, SemanticModel};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for access to the first or last element of `str.split()` without
+/// `maxsplit=1`
+///
+/// ## Why is this bad?
+/// Calling `str.split()` without `maxsplit` set splits on every delimiter in the
+/// string. When accessing only the first or last element of the result, it
+/// would be more efficient to only split once.
+///
+/// ## Example
+/// ```python
+/// url = "www.example.com"
+/// prefix = url.split(".")[0]
+/// ```
+///
+/// Use instead:
+/// ```python
+/// url = "www.example.com"
+/// prefix = url.split(".", maxsplit=1)[0]
+/// ```
+
+#[derive(ViolationMetadata)]
+pub(crate) struct MissingMaxsplitArg;
+
+impl Violation for MissingMaxsplitArg {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "Accessing only the first or last element of `str.split()` without setting `maxsplit=1`"
+            .to_string()
+    }
+}
+
+fn is_string(expr: &Expr, semantic: &SemanticModel) -> bool {
+    if let Expr::Name(name) = expr {
+        let Some(binding_id) = semantic.only_binding(name) else {
+            return false;
+        };
+        let binding = semantic.binding(binding_id);
+
+        if !typing::is_string(binding, semantic) {
+            return false;
+        }
+        return true;
+    } else if let Some(binding_id) = semantic.lookup_attribute(expr) {
+        let binding = semantic.binding(binding_id);
+        if !typing::is_string(binding, semantic) {
+            return false;
+        }
+        return true;
+    } else if let Expr::StringLiteral(_) = expr {
+        return true;
+    }
+
+    false
+}
```

Comment:
```suggestion
fn is_string(expr: &Expr, semantic: &SemanticModel) -> bool {
    if let Expr::Name(name) = expr {
        semantic
            .only_binding(name)
            .is_some_and(|binding_id| typing::is_string(semantic.binding(binding_id), semantic))
    } else if let Some(binding_id) = semantic.lookup_attribute(expr) {
        typing::is_string(semantic.binding(binding_id), semantic)
    } else {
        expr.is_string_literal_expr()
    }
}
```

Something like this might be a little neater.

Reply:
It absolutely is. The original looks downright goofy now that I'm looking back on it.